src/AMSlib/wf/basedb.hpp

```diff
@@ -1700,9 +1822,74 @@ class RMQPublisherHandler : public AMQP::LibEventHandler
     std::future_status status;
     while ((status = future.wait_for(span)) == std::future_status::timeout &&
            (iters++ < repeat))
-      std::future<ConnectionStatus> established;
+      std::future<RMQConnectionStatus> established;
     return status == std::future_status::ready;
   }
+
+  /**
+   *  @brief  Free the data pointed pointer in a vector and update vector.
+   *  @param[in]  addr            Address of memory to free.
+   *  @param[in]  buffer          The vector containing memory buffers
+   */
+  void free_ams_message(int msg_id, std::vector<AMSMessage>& buf)
+  {
+    const std::lock_guard<std::mutex> lock(ptr_mutex);
+    auto it =
+        std::find_if(buf.begin(), buf.end(), [&msg_id](const AMSMessage& obj) {
+          return obj.id() == msg_id;
+        });
+    if (it == buf.end()) {
+      WARNING(RMQPublisherHandler,
+              "Failed to deallocate msg #%d: not found",
+              msg_id)
+      return;
+    }
+    auto& msg = *it;
+    auto& rm = ams::ResourceManager::getInstance();
+    try {
+      rm.deallocate(msg.data(), AMSResourceType::HOST);
+    } catch (const umpire::util::Exception& e) {
+      WARNING(RMQPublisherHandler,
+              "Failed to deallocate #%d (%p)",
+              msg.id(),
+              msg.data());
+    }
+    DBG(RMQPublisherHandler, "Deallocated msg #%d (%p)", msg.id(), msg.data())
+    it = std::remove_if(buf.begin(),
```

Comment:
I am guessing remove_if will search again in buf. Didn't we search earlier?

Reply:
Fixed, thanks.